From 9a473089b1267048ad9c33f36f8e9856b320841a Mon Sep 17 00:00:00 2001
From: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
Date: Fri, 7 Nov 2014 10:28:15 +0000
Subject: [PATCH 283/351] mfd: arizona: Disable interrupts while suspending

An interrupt during suspend will cause a read of the status
register and the bus driver may not be available at this time.

Change-Id: I905bc157e9f7bdfe51539c95c8c98cabdc5d7af8
Signed-off-by: JS Park <aitdark.park@samsung.com>
Signed-off-by: Nariman Poushin <nariman@opensource.wolfsonmicro.com>
---
 drivers/mfd/arizona-core.c |   20 +++++++++++++-------
 1 file changed, 13 insertions(+), 7 deletions(-)

diff --git a/drivers/mfd/arizona-core.c b/drivers/mfd/arizona-core.c
index b6ca0b9..ed6232e 100644
--- a/drivers/mfd/arizona-core.c
+++ b/drivers/mfd/arizona-core.c
@@ -656,22 +656,28 @@ static int arizona_runtime_suspend(struct device *dev)
 #endif
 
 #ifdef CONFIG_PM_SLEEP
-static int arizona_suspend(struct device *dev)
+static int arizona_suspend_noirq(struct device *dev)
 {
 	struct arizona *arizona = dev_get_drvdata(dev);
 
-	dev_dbg(arizona->dev, "Suspend, disabling IRQ\n");
-	disable_irq(arizona->irq);
+	dev_dbg(arizona->dev, "Late suspend, reenabling IRQ\n");
+
+	if (arizona->irq_sem) {
+		enable_irq(arizona->irq);
+		arizona->irq_sem = 0;
+	}
 
 	return 0;
 }
 
-static int arizona_suspend_late(struct device *dev)
+static int arizona_suspend(struct device *dev)
 {
 	struct arizona *arizona = dev_get_drvdata(dev);
 
-	dev_dbg(arizona->dev, "Late suspend, reenabling IRQ\n");
-	enable_irq(arizona->irq);
+	dev_dbg(arizona->dev, "Early suspend, disabling IRQ\n");
+
+	disable_irq(arizona->irq);
+	arizona->irq_sem = 1;
 
 	return 0;
 }
@@ -708,7 +714,7 @@ const struct dev_pm_ops arizona_pm_ops = {
 			   NULL)
 	SET_SYSTEM_SLEEP_PM_OPS(arizona_suspend, arizona_resume)
 #ifdef CONFIG_PM_SLEEP
-	.suspend_late = arizona_suspend_late,
+	.suspend_noirq = arizona_suspend_noirq,
 	.resume_noirq = arizona_resume_noirq,
 #endif
 };
-- 
1.7.9.5

